components/places/src/bookmark_sync/store.rs

```diff
@@ -129,87 +136,398 @@ impl<'a> BookmarksStore<'a> {
+        self.apply_remote_items(now)?;
+
+        // Fires the `applyNewLocalStructure` trigger.
+        log::debug!("Applying new local structure");
+        self.interruptee.err_if_interrupted()?;
+        self.db
+            .execute_batch("DELETE FROM applyNewLocalStructureOps")?;
+
+        log::debug!("Resetting change counters for items that shouldn't be uploaded");
+        sql_support::each_chunk_mapped(
+            &ops.set_local_merged,
+            |op| op.merged_node.guid.as_str(),
+            |chunk, _| -> Result<()> {
+                self.interruptee.err_if_interrupted()?;
+                self.db.execute(
+                    &format!(
+                        "UPDATE moz_bookmarks SET
+                             syncChangeCounter = 0
+                         WHERE guid IN ({})",
+                        sql_support::repeat_sql_vars(chunk.len()),
+                    ),
+                    chunk,
+                )?;
+                Ok(())
+            },
+        )?;
+
+        log::debug!("Bumping change counters for items that should be uploaded");
+        sql_support::each_chunk_mapped(
+            &ops.set_local_unmerged,
+            |op| op.merged_node.guid.as_str(),
+            |chunk, _| -> Result<()> {
+                self.interruptee.err_if_interrupted()?;
+                self.db.execute(
+                    &format!(
+                        "UPDATE moz_bookmarks SET
+                             syncChangeCounter = 1
+                         WHERE guid IN ({})",
+                        sql_support::repeat_sql_vars(chunk.len()),
+                    ),
+                    chunk,
+                )?;
+                Ok(())
+            },
+        )?;
+
+        log::debug!("Flagging applied remote items as merged");
+        sql_support::each_chunk_mapped(
+            &ops.set_remote_merged,
+            |op| op.guid().as_str(),
+            |chunk, _| -> Result<()> {
+                self.interruptee.err_if_interrupted()?;
+                self.db.execute(
+                    &format!(
+                        "UPDATE moz_bookmarks_synced SET
+                             needsMerge = 0
+                         WHERE guid IN ({})",
+                        sql_support::repeat_sql_vars(chunk.len()),
+                    ),
+                    chunk,
+                )?;
+                Ok(())
+            },
+        )?;
+
+        Ok(())
+    }
 
-        // `itemsToMerge` is a view, so "deleting" from it fires the
-        // `insertNewLocalItems` and `updateExistingLocalItems`
-        // triggers instead.
-        self.db.execute_batch("DELETE FROM itemsToMerge")?;
+    fn apply_remote_items(&self, now: Timestamp) -> Result<()> {
+        log::debug!("Removing old tags");
+        self.interruptee.err_if_interrupted()?;
+        self.db.execute_batch(
+            "DELETE FROM moz_tags_relation
+             WHERE place_id IN (SELECT oldPlaceId FROM itemsToApply
```

Comment:
Double-check that this query can still use the `oldPlaceIds` and `newPlaceIds` indexes.

Also, verify that this does the right thing if we re-assign a synced tag to a new URL. Do we need a `WHERE tag_id IN (...)` clause, too?

Reply:
It can still use the indexes, and no, we don't want the `WHERE` clause. 😄